### DISCUSSION TITLE
Programmatically editing buffer breaks backspace and arrow keys

### DISCUSSION THREAD

**Discussion by jfmontanaro** [xonsh/xonsh#5263] [Q&A]
I'm trying to integrate [atuin](https://github.com/atuinsh/atuin) with xonsh. Right now I'm working on the actual search operation itself, which has two parts:

1. Run the actual search via `atuin`, which is a interactive TUI, and
2. If the search comes back with a result, put that in the buffer so that it can be edited/executed.

My problem is with step 2. I've got access to the buffer via methods similar to what's detailed on [this](https://xon.sh/tutorial_ptk.html) page, i.e. in my handler it's accessible as `event.current_buffer`. The problem is that when I edit the contents of the buffer, either by assigning directly to `buffer.text` or calling `buffer.set_document` (or `buffer.apply_completion`) it ends up in a weird broken state where I can type to add new text to the buffer, but I can't erase the text that I added programmatically. Navigating around with arrow keys is also broken and just results in spewing `^[[D` and the like to stdout.

What's going on here? How can I restore the buffer to a properly-editable state?

For reference, here's a simplified version of my code:
```xonsh
from prompt_toolkit.keys import Keys

def _atuin_search(buffer):
    with tempfile.NamedTemporaryFile('w+', encoding='ascii') as f:
        # atuin returns search result on stderr
        p = ![atuin search --interactive -- @(buffer.text) e>@(f.name)]
        search_result = f.read().strip()
        buffer.text = search_result
        buffer.cursor_position = len(search_result)
        
@events.on_ptk_create
def custom_keybindings(bindings, **kw):
    @bindings.add(Keys.ControlR)
    def search(event):
        _atuin_search(event.current_buffer)
```

**Comment by anki-code** [comment]
Some thoughts.

0. You can find a couple of issues about using fzf for the same goal. May be it helps.
1. I used subprocess module to workaround some issues in xontrib-argcomplete.
2. May be it helps to create your own completer like in xontrib-output-search.

If your common approach is working for ptk with pure python but not working in xonsh and you have strong spirit it will be cool to improve xonsh because of many issues around using completers like fzf.

_(My messages has "Maintainer" label but the "maintainer" is the team of volunteers during a couple of years. Don't treat my comment as a comment from "owner".)_

**Comment by jfmontanaro** [comment]
Thanks! Your suggestion about `fzf` led me to [xontrib-fzf-widgets](laloch/xontrib-fzf-widgets), which didn't seem to suffer from the same problem. I saw that it was just using `subprocess.Popen` directly to run `fzf`, so I switched my solution to do the same and voila, the issue went away. 

Previously I was running the command via `![]` and redirecting stderr to a temporary file, because it uses stdin/stdout to talk to the user and returns the search result on stderr. Not sure what xonsh is doing differently from raw `subprocess`, but it works so I'm not asking too many questions.

~~This does introduce another minor problem, though. Commands run via `subprocess` directly don't see updates made to the environment from within xonsh, unless `$UPDATE_OS_ENVIRON` is `True`. I could just set `$UPDATE_OS_ENVIRON` as part of the integration, but I'm hesitant to do that as it seems like leaking an implementation detail.~~

~~However I do need the xonsh environment available to the `atuin` command, in case e.g. the user has set one of the `$XDG_*` values that change where config/data get stored, or something like that.~~

~~Does xonsh expose any kind of convert-xonsh-env-to-dict logic anywhere that I could use, and then just pass that as the env explicitly?~~

~~Another option would be to iterate over `${...}` and call `get_stringified()` on each key, but I'm not the biggest fan of that solution either because not everything in the xonsh environment should be passed to subprocesses. Obviously I can just ignore exceptions for values that can't be stringified (e.g. functions), but I don't know if that will exclude everything that should be excluded. I'd prefer to use the built-in logic if available.~~

Edit: Scratch all that! I just discovered `${...}.detype()` which is exactly what I need.
